### PR TITLE
Update `sentry_dio` to dio 5

### PR DESCRIPTION
**Pull request by ueman**
## :scroll: Description

Dio changed in ownership to the guys behind diox, which previously was forked from dio.
Since the ownership changed, they've published diox v5 also as dio v5.
This PR makes the necessary changes to be compatible.

## :bulb: Motivation and Context
Compatibility with dio v5

## :green_heart: How did you test it?
Tests

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPii` is enabled
- [x] I updated the docs if needed
- [x] All tests passing
- [ ] No breaking changes

### CODE REVIEW

**Comment by krystofwoldrich** [comment]
Thanks, it looks good, but CI is not happy yet. :)

**Comment by kuhnroyal** [comment]
There are a few things missing.

**Comment by ueman** [comment]
> Thanks, it looks good, but CI is not happy yet. :)

I'll look at it later.

> There are a few things missing.

What did I miss?

**Comment by kuhnroyal** [comment]
> > Thanks, it looks good, but CI is not happy yet. :)
> 
> I'll look at it later.
> 
> > There are a few things missing.
> 
> What did I miss?

Nah all good, sorry. I was still working on the 6.x branch.

**Comment by kuhnroyal** [comment]
Would be great if we can get this in a new 7.0.0 beta asap.

**Comment by ueman** [comment]
> Would be great if we can get this in a new 7.0.0 beta asap.

I'm on it 🚀

File: dio/test/breadcrumb_client_adapter_test.dart [hunk 1]
```diff
@@ -98,7 +98,7 @@ void main() {
         await sut.get<dynamic>('/');
         fail('Method did not throw');
       } on DioError catch (e) {
-        expect(e.message, 'Exception: test');
+        expect(e.error.toString(), 'Exception: test');
```
Comment (ueman) [inline dio/test/breadcrumb_client_adapter_test.dart:101]: This is due to changes in dio

File: dio/test/dio_event_processor_test.dart [hunk 2]
```diff
@@ -75,7 +75,6 @@ void main() {
       expect(processedEvent.request?.queryString, 'foo=bar');
       expect(processedEvent.request?.headers, <String, String>{
         'foo': 'bar',
-        'content-type': 'application/json; charset=utf-8'
```
Comment (ueman) [inline dio/test/dio_event_processor_test.dart:78]: dio doesn't apply a content-type by itself anymore

**Comment by kuhnroyal** [comment]
The `dio` overrides in the example are completely messed up.

File: flutter/example/pubspec.yaml [hunk 3]
```diff
@@ -18,7 +18,7 @@ dependencies:
   universal_platform: ^1.0.0
   feedback: ^2.0.0
   provider: ^6.0.0
-  dio: ^4.0.0
+  dio: any # This gets constrained by `sentry_dio`
```
Comment (kuhnroyal) [inline flutter/example/pubspec.yaml:21]: Same needs to be done for the min-version-test

**Review by brustolin — APPROVED** [review]
LGTM

**Review by krystofwoldrich — APPROVED** [review]
LGTM 🚀

**Comment by ueman** [comment]
Is there something left to do before it can be merged?

**Review by kuhnroyal — APPROVED** [review]

**Comment by krystofwoldrich** [comment]
@ueman I think it's ready, just let's update the changelog to show this as a breaking change.

File: CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,7 @@
 ### Fixes
 
 - Set client name with version in Android SDK ([#1274](https://github.com/getsentry/sentry-dart/pull/1274))
+- Update `sentry_dio` to dio v5 ([#1282](https://github.com/getsentry/sentry-dart/pull/1282))
```
Comment (ueman) [inline CHANGELOG.md]:
```suggestion

### Breaking Changes

- Update `sentry_dio` to dio v5 ([#1282](https://github.com/getsentry/sentry-dart/pull/1282))
```

**Comment by kuhnroyal** [comment]
Would love to get a beta release with this :)

**Comment by ueman** [comment]
The pipeline still has one failing job, but it's for the pure Dart SDK. Since that's not touch in this PR, I'm wondering whether that's just a flake. Can someone rerun the jobs?

Edit: Apparently, that job is failing for all PRs which target the v7 branch, so it's unrelated to this PR.

**Comment by marandaneto** [comment]
@ueman 

> The `build_web_compilers` version – 3.2.7 – is not within the allowed constraint – ^4.0.0.

We have to upgrade `build_web_compilers` to `^4.0.0` and `build_runner` to `^2.4.0` cus the beta channel uses Dart 3 and the new versions require Dart 3.
Pretty much the inverse of https://github.com/getsentry/sentry-dart/pull/1299/
Can be another PR since its not related to this issue.

**Comment by marandaneto** [comment]
I will look into this tomorrow, sorry about the delay, just came back from vacation and catching up with all the pending things.

**Comment by ueman** [comment]
No worries, thanks for looking into it.

**Review by marandaneto — APPROVED** [review]
Thanks for doing this @ueman

**Comment by marandaneto** [comment]
Mind appending the [migration page](https://docs.sentry.io/platforms/flutter/migration/#migrating-from-sentry_flutter-618x-to-sentry-700)?

**Comment by ueman** [comment]
> Mind appending the [migration page](https://docs.sentry.io/platforms/flutter/migration/#migrating-from-sentry_flutter-618x-to-sentry-700)?

Sure. Is there anything in particular you'd like to see?

Edit: It's here https://github.com/getsentry/sentry-docs/pull/6400

**Comment by kuhnroyal** [comment]
I have tested this change excessively over the last few days and the whole update from Sentry 6.x/Dio 4.x to Sentry 7.0/Dio 5.0 is a big step back regarding traceability of the `DioError`. This is in big part due to changes in `dio` which removed the whole source stacktrace. But due to the `DioErrorExtractor` there are now also cases where the inner stacktrace of `DioError.error.stacktrace` is completely missing in the event.

I think the `DioError.stacktrace` needs to be added by default, as it was the case in `6.x` and only the inner cause error/stacktrace should be extracted via the `DioErrorExtractor`.

Something like:
```dart
class DioErrorExtractor extends ExceptionCauseExtractor<DioError> {
  @override
  ExceptionCause? cause(DioError error) {
    if (error.error == null) {
      return null;
    }
    return ExceptionCause(
      error.error,
      error.error is Error ? (error.error as Error).stackTrace : null,
    );
  }
}
```

**Pull request merged**